test/jdk/sun/security/pkcs11/SecretKeyFactory/TestPBKD.java

```diff
@@ -0,0 +1,515 @@
+        BigInteger staticExpectedKey = new BigInteger(staticExpectedKeyString,
+                16);
+        BigInteger expectedKey;
+        try {
+            // Since we need to access an internal
+            // SunJCE API, we use reflection.
+            Class<?> PKCS12PBECipherCore = Class.forName(
+                    "com.sun.crypto.provider.PKCS12PBECipherCore");
+
+            Field macKeyField = PKCS12PBECipherCore.getDeclaredField("MAC_KEY");
+            macKeyField.setAccessible(true);
+            int MAC_KEY = (int) macKeyField.get(null);
+
+            Method deriveMethod = PKCS12PBECipherCore.getDeclaredMethod(
+                    "derive", char[].class, byte[].class, int.class,
+                    int.class, int.class, String.class, int.class);
+            deriveMethod.setAccessible(true);
+            expectedKey = i((byte[]) deriveMethod.invoke(null,
+                    keySpec.getPassword(), keySpec.getSalt(),
+                    keySpec.getIterationCount(), keySpec.getKeyLength() / 8,
+                    MAC_KEY, hashAlgo, blockLen));
+            checkAssertionValues(expectedKey, staticExpectedKey);
+        } catch (ReflectiveOperationException ignored) {
+            expectedKey = staticExpectedKey;
+        }
+        return new AssertionData(algo, keySpec, expectedKey);
+    }
+
+    private static AssertionData pbkd2AssertionData(String algo,
+            char[] password, int keyLen, String kdfAlgo,
+            String staticExpectedKeyString) {
+        PBEKeySpec keySpec = new PBEKeySpec(password, salt, iterations, keyLen);
+        BigInteger staticExpectedKey = new BigInteger(staticExpectedKeyString,
+                16);
+        BigInteger expectedKey = null;
+        if (sunJCE != null) {
+            try {
+                expectedKey = i(SecretKeyFactory.getInstance(kdfAlgo, sunJCE)
+                        .generateSecret(keySpec).getEncoded());
+                checkAssertionValues(expectedKey, staticExpectedKey);
+            } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
+                // Move to staticExpectedKey as it's unlikely
+                // that any of the algorithms are available.
+                sunJCE = null;
+            }
+        }
+        if (expectedKey == null) {
+            expectedKey = staticExpectedKey;
+        }
+        return new AssertionData(algo, keySpec, expectedKey);
+    }
+
+    private static void checkAssertionValues(BigInteger expectedValue,
+            BigInteger staticExpectedValue) {
+        if (!expectedValue.equals(staticExpectedValue)) {
+            printHex("SunJCE value", expectedValue);
+            printHex("Static value", staticExpectedValue);
+            throw new Error("Static and SunJCE values do not match.");
+        }
+    }
+
+    private static final char[] pwd = "123456\uA4F7".toCharArray();
+    private static final char[] emptyPwd = new char[0];
+    private static final byte[] salt = "abcdefgh".getBytes();
```

Comment:
My other comment of charset encoding applies here as well?

Reply:
Good. Fixed for salt here.